### PR TITLE
[libc++][span] P2821R5: `span.at()`

### PR DESCRIPTION
**Pull request by H-G-Hristov**
- Implements: [P2821R5: span.at()](https://wg21.link/P2821R5)
  (https://eel.is/c++draft/views.contiguous#views.span)
- Cleaned up `span.operator[]` test

### CODE REVIEW

File: libcxx/include/span [hunk 1]
```diff
@@ -343,6 +347,15 @@ public:
         return __data_[__idx];
     }
 
+#  if _LIBCPP_STD_VER >= 26
+    _LIBCPP_HIDE_FROM_ABI constexpr reference at(size_type __idx) const {
+      if (__idx >= size()) {
+        __throw_out_of_range();
+      }
+      return *(data() + __idx);
```
Comment (philnik777) [inline libcxx/include/span:355]:
```suggestion
      if (__idx >= size())
        __throw_out_of_range();
      return data()[__idx];
```
Maybe also rename `__idx` to `__index` for better readability. Same below

File: libcxx/test/std/containers/views/views.span/span.elem/op_idx.pass.cpp [hunk 2]
```diff
@@ -41,7 +41,6 @@ void testRuntimeSpan(Span sp, std::size_t idx)
     assert(r1 == r2);
 }
 
-struct A{};
```
Comment (philnik777) [inline libcxx/test/std/containers/views/views.span/span.elem/op_idx.pass.cpp:44]: Unrelated change?

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+void test_exceptions() {
+#ifndef TEST_HAS_NO_EXCEPTIONS
+  // With static extent
+  {
+    std::array arr{1, 2, 3, 4};
+    const std::span arrSpan{arr};
+
+    try {
+      TEST_IGNORE_NODISCARD arrSpan.at(arr.size() + 1);
```
Comment (philnik777) [inline libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]:
```suggestion
      (void) arrSpan.at(arr.size() + 1);
```
I'm pretty sure the argument should be `arr.size()`, since that's the first out of bounds element.

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    testSpanAt(vecSpan, 0, 0);
+    testSpanAt(vecSpan, 1, 1);
+    testSpanAt(vecSpan, 6, 9084);
+  }
+
+  return true;
+}
+
+void test_exceptions() {
```
Comment (philnik777) [inline libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:67]: I'd also like to see tests with the last element in range and something like `numeric_limits<size_type>::max()` to have something not even close covered.

**Review by philnik777 — CHANGES_REQUESTED** [review]
Looks mostly good, but I have a few small comments.

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for working in this.

I didn't start the CI jobs since I know they will fail.

**Review by mordante — COMMENTED** [review]
I pressed submit too soon ;-)

You need to update the FTM as implemented too and update the generated files.

File: libcxx/include/span (hunk 1 above)
Comment (philnik777) [reply libcxx/include/span:355]: Honestly, that seems really weird to me. Why would you write in in such a convoluted way? This seems like a case where the wording should simply be adjusted to something simpler, or I'm missing something and there is a test missing.

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp (hunk 3 above)
Comment (philnik777) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: For what reason though? In which case is `TEST_IGNORE_NODISCARD` better to understand than `(void)`? I'm pretty sure I've pushed back against uses of this macro in the past and haven't gotten any good response for why one would ever use it. IMO this makes things just more confusing.

File: libcxx/test/std/containers/views/views.span/span.elem/op_idx.pass.cpp (hunk 2 above)
Comment (H-G-Hristov) [reply libcxx/test/std/containers/views/views.span/span.elem/op_idx.pass.cpp:44]: Yes, I can restore that change. Initially I was looking into reusing op_idx tests and I noticed that this declaration is unused.

File: libcxx/include/span (hunk 1 above)
Comment (H-G-Hristov) [reply libcxx/include/span:355]: I already renamed the arg as you suggested but in the synopsis both methods would be using `idx`
```c++
constexpr reference operator[](size_type idx) const;
constexpr reference at(size_type idx) const;
```
Would it make sense to also rename the arg of `operator[]` for consistency?

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp (hunk 3 above)
Comment (H-G-Hristov) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: C++ Core Guidelines: [ES.48: Avoid casts](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es48-avoid-casts)

> Never cast to (void) to ignore a [[nodiscard]]return value. If you deliberately want to discard such a result, first think hard about whether that is really a good idea (there is usually a good reason the author of the function or of the return type used [[nodiscard]] in the first place). If you still think it's appropriate and your code reviewer agrees, use std::ignore = to turn off the warning which is simple, portable, and easy to grep.
Alternatives
...
>    - Use std::ignore = to ignore [[nodiscard]] values.

Maybe use `std::ignore` instead?

Comment (philnik777) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: Sure, why not.

File: libcxx/include/span (hunk 1 above)
Comment (H-G-Hristov) [reply libcxx/include/span:355]: For consistency I changed the implementation to how it is implemented in the other methods. Would that be a good compromise?

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp (hunk 4 above)
Comment (H-G-Hristov) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: I used `numeric_limits` I hope my change isn't too far off what you mean.

**Comment by H-G-Hristov** [comment]
I don't know what is the accepted protocol, who should click  the "Resolve conversation" buttons (e.g. the author...), so I'll leave them unclicked.
Thank you for the review!

File: libcxx/include/span [hunk 5]
```diff
@@ -343,6 +345,15 @@ public:
         return __data_[__idx];
     }
 
+#  if _LIBCPP_STD_VER >= 26
+    _LIBCPP_HIDE_FROM_ABI constexpr reference at(size_type __index) const {
+      if (__index >= size()) {
```
[the remark this answers is not in the document]
Comment (H-G-Hristov) [reply libcxx/include/span]: Done! Sorry I missed that.

File: libcxx/include/span [hunk 6]
```diff
@@ -343,6 +347,15 @@ public:
         return __data_[__idx];
     }
 
+#  if _LIBCPP_STD_VER >= 26
+    _LIBCPP_HIDE_FROM_ABI constexpr reference at(size_type __idx) const {
```
[the remark this answers is not in the document]
Comment (mordante) [reply libcxx/include/span]: Nevermind this only applies to non-member functions and classes themselves.

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    } catch (...) {
+      assert(false);
+    }
+
+    try {
+      std::ignore = arrSpan.at(arr.size() - 1);
+      // pass
+      assert(arrSpan.at(arr.size() - 1) == std::numeric_limits<int>::max());
+    } catch (const std::out_of_range&) {
+      assert(false);
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:99]:
```suggestion
```
It think it looks better to only use a `catch(...)` when we don't expect an exception.

File: z_libcxx-dev-tools
Comment (mordante) [inline z_libcxx-dev-tools:1]: What is this submodule? I assume it's not intended to be here.

**Review by mordante — CHANGES_REQUESTED** [review]
Mostly looks good, a few comments.

File: z_libcxx-dev-tools
Comment (H-G-Hristov) [reply z_libcxx-dev-tools:1]: Thank you for noticing. Somehow what should have been ignored snuck in back again.

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  {
+    std::array<int, 0> arr{};
+    const std::span arrSpan{arr};
+
+    try {
+      std::ignore = arrSpan.at(0);
+      assert(false);
+    } catch (const std::out_of_range& e) {
+      // pass
+      assert(e.what() == "span"s);
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:112]:
```suggestion
      LIBCPP_ASSERT(e.what() == "span"s);
```
This assert should only be tested for libc++. This means on other implementations the `e` is unused.

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    } catch (...) {
+      assert(false);
+    }
+
+    try {
+      std::ignore = arrSpan.at(arr.size() - 1);
+      // pass
+      assert(arrSpan.at(arr.size() - 1) == std::numeric_limits<int>::max());
+    } catch (...) {
```
Comment (mordante) [inline libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:98]: please apply this change to similar places too,

**Review by mordante — COMMENTED** [review]
A few more comments, I overlooked parts of them yesterday.

Please make sure to apply the suggestions to all similar places in the code.

**Review by mordante — APPROVED** [review]
Thanks for the fixes, LGTM!

**Comment by H-G-Hristov** [comment]
> Thanks for the fixes, LGTM!

Thank you!

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp (hunk 4 above)
Comment (philnik777) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: I meant that you should test `span.at(numeric_limits<size_type>::max())`.

**Comment by H-G-Hristov** [comment]
@philnik777  a gentle ping 
...and Happy New Year!

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp (hunk 4 above)
Comment (philnik777) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: I don't think you've addressed this?

**Review by philnik777 — APPROVED** [review]
Other than the test this LGTM.

File: libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp (hunk 4 above)
Comment (H-G-Hristov) [reply libcxx/test/std/containers/views/views.span/span.elem/at.pass.cpp:75]: Sorry, I missed your comment!

**Comment by Zingam** [comment]
Thank you very much for the review. I'll wait a few days for addional feedback before I land this PR.
